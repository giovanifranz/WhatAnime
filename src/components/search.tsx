'use client'

import { useCallback } from 'react'
import { useForm } from 'react-hook-form'
import { FiSearch } from 'react-icons/fi'
import UseAnimations from 'react-useanimations'
import loading from 'react-useanimations/lib/loading'

import { zodResolver } from '@hookform/resolvers/zod'
import * as AccessibleIcon from '@radix-ui/react-accessible-icon'
import { z } from 'zod'

import { animeStore } from '@/store/animeStore'

import { Button } from './ui/button'
import {
  FormField,
  FormItem,
  FormLabel,
  FormControl,
  Form,
  FormDescription,
  FormMessage,
} from './ui/form'
import { Input } from './ui/input'
/*import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from './ui/select'*/

const formSchema = z.object({
  title: z.string().min(3).max(30),
  select: z.enum(['word', 'upload']),
})

export function Search() {
  const { getAnimesByTitle, isLoading, error } = animeStore((state) => ({
    getAnimesByTitle: state.getAnimesByTitle,
    isLoading: state.byTitle?.isLoading || false,
    error: state.byTitle?.error || null,
  }))
  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      title: '',
      select: 'word',
    },
  })

  const onSubmit = useCallback(
    (values: z.infer<typeof formSchema>) => {
      getAnimesByTitle(values.title)
    },
    [getAnimesByTitle],
  )

  return (
    <Form {...form}>
      <form
        onSubmit={form.handleSubmit(onSubmit)}
        className="flex h-48 max-w-4xl flex-col justify-center gap-4"
      >
        {
          <FormField
            control={form.control}
            name="select"
            render={
              (/*{ field }*/) => (
                <FormItem>
                  <div className="flex h-12 items-center gap-4">
                    <FormLabel className="scroll-m-20 text-2xl font-semibold uppercase tracking-tight md:text-3xl">
                      Search
                    </FormLabel>

                    {/*<FormControl>
                  <Select onValueChange={field.onChange} defaultValue={field.value}>
                    <SelectTrigger className="w-24">
                      <SelectValue placeholder="Word" />
                    </SelectTrigger>
                    <SelectContent>
                      <SelectGroup>
                        <SelectItem value="word">Word</SelectItem>
                        <SelectItem value="upload">Upload</SelectItem>
                      </SelectGroup>
                    </SelectContent>
                  </Select>
          </FormControl>*/}
                  </div>
                </FormItem>
              )
            }
          />
        }
        <div className="flex w-full max-w-full items-center space-x-2">
          <FormField
            control={form.control}
            name="title"
            render={({ field }) => (
              <FormItem className="w-full">
                <FormControl>
                  <Input
                    type="text"
                    minLength={3}
                    maxLength={30}
                    placeholder="Enter your search key word"
                    {...field}
                  />
                </FormControl>
              </FormItem>
            )}
          />
          <Button type="submit" className="w-16" disabled={isLoading}>
            <AccessibleIcon.Root label="Search Button">
              {isLoading ? <UseAnimations animation={loading} /> : <FiSearch size={20} />}
            </AccessibleIcon.Root>
          </Button>
        </div>
        {!error ? (
          <FormDescription className="capitalize">
            Search your favorite Anime
          </FormDescription>
        ) : (
          <FormMessage className="capitalize">{error}</FormMessage>
        )}
      </form>
    </Form>
  )
}
