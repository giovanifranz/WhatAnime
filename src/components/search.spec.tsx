import { render, screen } from '@/test/test-utils'

import { Search } from './search'

describe('Teste UnitÃ¡rio - Search', () => {
  it('Deve renderizar o Search corretamente', async () => {
    const { asFragment } = render(<Search />)

    expect(asFragment()).toMatchSnapshot()

    const inputName = screen.getByPlaceholderText(/Enter your search key word/i)

    expect(inputName).toBeVisible()
  })
})
