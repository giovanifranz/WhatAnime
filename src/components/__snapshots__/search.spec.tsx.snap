// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Teste UnitÃ¡rio - Search > Deve renderizar o Search corretamente 1`] = `
<DocumentFragment>
  <script>
    !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&false)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}else{c.add('dark')}if(e==='light'||e==='dark'||!e)d.style.colorScheme=e||'dark'}catch(e){}}()
  </script>
  <form
    class="flex h-48 max-w-4xl flex-col justify-center gap-4"
  >
    <div
      class="space-y-2"
    >
      <div
        class="flex h-12 items-center gap-4"
      >
        <label
          class="peer-disabled:cursor-not-allowed peer-disabled:opacity-70 scroll-m-20 text-2xl font-semibold uppercase tracking-tight md:text-3xl"
          for=":r0:-form-item"
        >
          Search
        </label>
      </div>
    </div>
    <div
      class="flex w-full max-w-full items-center space-x-2"
    >
      <div
        class="space-y-2 w-full"
      >
        <input
          aria-describedby=":r1:-form-item-description"
          aria-invalid="false"
          class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
          id=":r1:-form-item"
          maxlength="30"
          minlength="3"
          name="title"
          placeholder="Enter your search key word"
          type="text"
          value=""
        />
      </div>
      <button
        class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-16"
        type="submit"
      >
        <svg
          aria-hidden="true"
          fill="none"
          focusable="false"
          height="20"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
        <span
          style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
        >
          Search Button
        </span>
      </button>
    </div>
    <p
      class="text-sm text-muted-foreground capitalize"
      id="undefined-form-item-description"
    >
      Search your favorite Anime
    </p>
  </form>
</DocumentFragment>
`;
